vernac/comProgramFixpoint.ml

```diff
@@ -160,48 +162,52 @@ let build_wellfounded pm (recname,pl,bl,arityc,body) ?scope ?clearbody poly ?typ
   let binders = Evarutil.nf_rel_context_evar sigma binders in
   let top_arity = Evarutil.nf_evar sigma top_arity in
   let make = Evarutil.nf_evar sigma make in
-  let hook, recname, typ =
+  let recname_func, typ =
     if List.length binders_rel > 1 then
-      let name = add_suffix recname "_func" in
-      let hook { Declare.Hook.S.dref; uctx; _ } =
+      add_suffix recname "_func", it_mkProd_or_LetIn top_arity binders
+    else
+      recname, it_mkProd_or_LetIn top_arity binders_rel in
+  RetrieveObl.check_evars env sigma;
+  let evars, (_, evmap), evars_def, evars_typ =
+    RetrieveObl.retrieve_obligations env recname_func sigma 0 def typ
+  in
+  let hook =
+    if List.length binders_rel > 1 then
+      let hook { Declare.Hook.S.dref; uctx; obls; _ } =
+        let update c = CVars.replace_vars obls (evmap mkVar (Evarutil.nf_evar (Evd.from_ctx uctx) c)) in
+        let make = update make in
+        let top_arity = update top_arity in
+        let binders_rel = Context.Rel.map_het update binders_rel in
         let univs = UState.check_univ_decl ~poly uctx udecl in
         let h_body =
           let inst = UState.(match fst univs with
-              | Polymorphic_entry uctx -> EConstr.EInstance.make (UVars.UContext.instance uctx)
-              | Monomorphic_entry _ -> EConstr.EInstance.empty) in
-          EConstr.mkRef (dref, inst) in
-        let body = it_mkLambda_or_LetIn (mkApp (h_body, [|make|])) binders_rel in
-        let ty = it_mkProd_or_LetIn top_arity binders_rel in
-        let ty = EConstr.Unsafe.to_constr ty in
+              | Polymorphic_entry uctx -> UVars.UContext.instance uctx
+              | Monomorphic_entry _ -> UVars.Instance.empty) in
+          Constr.mkRef (dref, inst) in
+        let body = Term.it_mkLambda_or_LetIn (Constr.mkApp (h_body, [|make|])) binders_rel in
         (*FIXME poly? *)
-        let sigma = Evd.from_ctx uctx in
-        let ce = definition_entry ~types:ty ~univs (EConstr.to_constr sigma body) in
+        let ty = Term.it_mkProd_or_LetIn top_arity binders_rel in
+        let ce = definition_entry ~types:ty ~univs body in
         (* FIXME: include locality *)
```

Comment:
I think this comment has been outdated since 5cbcc8fd761df0779f6202fef935f07cfef8a228 (2016)

Reply:
https://github.com/coq/coq/pull/18987